### PR TITLE
Limit the maximum bytes length of blocks

### LINKED ISSUE #201: Limit size of block (byte-wise)
Currently, Libplanet doesn't define how many transactions will be stored in a block, and the size of a transaction is also not explicitly defined. however, in order to predict figures such as storage space and network traffic, it is necessary to be able to determine how much physical space a block occupies.

The block size is not a protocol-level constraint, and game developers need to adjust the unit to synchronize the status of each game, so I suggest offering it as an configurable parameter.

### PR DESCRIPTION
**Pull request by dahlia**
This fixes #201.  Please read added changelogs for details.

### CODE REVIEW

**Review by earlbread — DISMISSED** [review]

**Comment by dahlia** [comment]
@planetarium/libplanet Please review this.

File: Libplanet/Blockchain/BlockChain.cs [hunk 1]
```diff
@@ -682,63 +685,120 @@ public long GetNextTxNonce(Address address)
...
+                index: index,
+                difficulty: difficulty,
+                totalDifficulty: Tip.TotalDifficulty,
+                nonce: default,
+                miner: miner,
+                previousHash: prevHash,
+                timestamp: currentTime,
+                transactions: new Transaction<T>[0]
+            ).BytesLength;
+            int maxBlockBytes = Math.Max(Policy.GetMaxBlockBytes(index), 1);
```
Comment (limebell) [inline Libplanet/Blockchain/BlockChain.cs:744]: What is the meaning of `1`?

**Review by riemannulus — DISMISSED** [review]

File: Libplanet/Blockchain/BlockChain.cs (hunk 1 above)
Comment (dahlia) [reply Libplanet/Blockchain/BlockChain.cs:744]: This makes the `maxBlockBytes` 1 at least even if `Policy.GetMaxBlockBytes()` is less than 1 (e.g., `Max(-1, 1)` → `1`, `Max(5, 1)` → `5`).

File: Libplanet/Blocks/InvalidBlockBytesLengthException.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                $"Allowed (max): #{maxBlockBytesLength}")
+        {
+            BlockBytesLength = blockBytesLength;
+            MaxBlockBytesLength = maxBlockBytesLength;
+        }
+
+        /// <summary>
+        /// The bytes length of the actual block.
+        /// </summary>
+        public int BlockBytesLength { get; set; }
```
Comment (longfin) [inline Libplanet/Blocks/InvalidBlockBytesLengthException.cs:41]: I guess we need to add codes about serialization for these fields to preserve them.

File: Libplanet/Blockchain/BlockChain.cs (hunk 1 above)
Comment (limebell) [reply Libplanet/Blockchain/BlockChain.cs:744]: I was wondering why the value is exactly 1 🤔

Comment (dahlia) [reply Libplanet/Blockchain/BlockChain.cs:744]: This implements the below documented behavior:

<https://github.com/planetarium/libplanet/blob/e7d963bae4a757f45ac808d679e31a060a5ed780/Libplanet/Blockchain/Policies/IBlockPolicy.cs#L81-L82>

which assumes there is no block taking 0 bytes or negative length of bytes.

**Review by limebell — DISMISSED** [review]

File: Libplanet/Blocks/InvalidBlockBytesLengthException.cs (hunk 2 above)
Comment (dahlia) [reply Libplanet/Blocks/InvalidBlockBytesLengthException.cs:49]: Added.

**Comment by dahlia** [comment]
@planetarium/libplanet Addressed reviews and rebased.  Please review this again!

**Review by longfin — DISMISSED** [review]

**Review by earlbread — DISMISSED** [review]

**Review by limebell — DISMISSED** [review]

**Review by riemannulus — DISMISSED** [review]

**Review by moreal — DISMISSED** [review]

**Comment by dahlia** [comment]
@planetarium/libplanet I rebased.  Please review this again.

**Review by earlbread — APPROVED** [review]

**Review by longfin — APPROVED** [review]

**Review by riemannulus — APPROVED** [review]

**Review by moreal — APPROVED** [review]

**Comment by dahlia** [comment]
@planetarium/libplanet As it seems to reach Azure Pipelines' rate limit, the build is being timed out again and again.  Instead retring the build again, I am going to merge this anyway.

**Pull request merged**